### ISSUE TITLE
atproto oauth2 vs. createSession, scope problems: (transition:generic vs. com.atproto.access)

### ISSUE THREAD

**Issue by reflectivist** [bluesky-social/atproto#3212]
### Discussed in https://github.com/bluesky-social/atproto/discussions/3211

<div type='discussions-op-text'>

<sup>Originally posted by **reflectivist** December  9, 2024</sup>
On iOS, using oauth2, I have custom code that gets authorization and fetches a bsky access token (it does the DPoP / PKCE / nonce stuff). When I decode the accessJwt, shows it has scope **"atproto transition:generic"**.

However, when I try to access **/xrpc/com.atproto.repo.uploadBlob** with that token, I get the error "Bad token scope".

However, if I access **/xrpc/com.atproto.server.createSession** endpoint from curl, and hardcode that access token into my iOS code, it works fine for uploading a blob and creating a record. The createSession accessJwt token's scope says **"com.atproto.access"**

Unfortunately if I explicitly include **com.atproto.access** along with the **"atproto transistion:generic"** scopes  in the oauth2 configureion as well as client-metadata, I no longer can fetch an access token from iOS app. Oauth2 flow only respects "atproto and transition:generic" scopes that I've seen.

I need to get uploadBlob to work with oauth based authentication, because I don't want to have my app handle user's login credentials directly, whereas createSession seems to require a password. I'm not sure how else to create a session and get the necessary scope authorized.

_I'm making it an issue because I've been digging around all over the place in the documentation, API but there's not much that clarifies any of it and I've read that transition:generic is supposed to enable pretty comprehensive API access._
</div>

**Comment by bobbyzholmes** [comment]
Yeah I'm getting the same error on API requests that seem to update data on bsky, fetching data seems fine. But using an OAuth JWT with the same scopes and making a `com.atproto.repo.createRecord` for instance causes a bad token scope to be returned vs the JWT from `createSession` the same requests work fine

```
{
    "error": "InvalidToken",
    "message": "Bad token scope"
}
```

**Comment by yamarten** [comment]
[bsky.social entryway](https://docs.bsky.app/docs/advanced-guides/entryway) works as a proxy, but it may not work properly for API requests with OAuth authentication. By the way, they are going to remove the proxy function in the near future.

https://docs.bsky.app/blog/2025-protocol-roadmap-spring#deprecations

Does the same error occur even if you send a request directly to PDS (*.host.bsky.network) using the same token?

**Comment by bobbyzholmes** [comment]
Hey Yamarten

Apologies for delayed reply, I presume you mean extracting the host from "aud" during the auth flow, I get the below response from: https://bsky.social/oauth/authorize/sign-in

```
{
  "account": {
    "sub": "did:REMOVED",
    "aud": "did:web:hollowfoot.us-west.host.bsky.network",
    "email": "REMOVED",
    "email_verified": true,
    "preferred_username": "REMOVED"
  },
  "consentRequired": false
}
```

And contained in the JWT
```
{
  "typ": "at+jwt",
  "alg": "ES256K"
}

{
  "aud": "did:web:hollowfoot.us-west.host.bsky.network",
  "iat": REMOVED,
  "exp": REMOVED,
  "sub": "REMOVED",
  "jti": "REMOVED",
  "cnf": {
    "jkt": "REMOVED"
  },
  "client_id": "https://REMOVED/dev-meta.json",
  "scope": "atproto transition:generic",
  "iss": "https://bsky.social"
}
```



Making the createRecord request to: `https://hollowfoot.us-west.host.bsky.network/xrpc/com.atproto.repo.createRecord`

Returns:

```
{
    "error": "InvalidToken",
    "message": "Malformed token"
}
```

vs

```
{
    "error": "InvalidToken",
    "message": "Bad token scope"
}
```

Sent to: `https://bsky.social/xrpc/com.atproto.repo.createRecord`

I get the same "Bad token scope" for requests to `https://bsky.social/xrpc/com.atproto.sync.getLatestCommit`


But the same token works for requests to the below endpoints:

`https://bsky.social/xrpc/com.atproto.repo.listRecords` or
`https://bsky.social/xrpc/com.atproto.repo.describeRepo`

**Comment by yamarten** [comment]
> I presume you mean extracting the host from "aud" during the auth flow

That's right. Another way is to [resolve](https://docs.bsky.app/docs/advanced-guides/resolving-identities) the ID and get the DID document. If you are trying to support users other than Bluesky PDS, ID resolution is also necessary to know the OAuth endpoint. 

> ```
> {
>     "error": "InvalidToken",
>     "message": "Malformed token"
> }
> ```

Please check if you have passed the access token as a bearer. If not, it seems difficult for me to be more useful.

> But the same token works for requests to the below endpoints:
>
> `https://bsky.social/xrpc/com.atproto.repo.listRecords` or
> `https://bsky.social/xrpc/com.atproto.repo.describeRepo`

These APIs do not validate the access token, so I think it is a reasonable result.

**Comment by bobbyzholmes** [comment]
Finally got back to reviewing this issue, and some helpful insight on Stack OverFlow.

Yes I will look to support other Bluesky PDS, but at this point just getting the OAuth working against bsky will help prove the functionality is working

What I had missed from the docs was the requirement to use, Dpop within the Authorization header format `Authorization: DPoP <token>` (not Bearer <token>)

This change finally got me pass the issue with the token, but I'm still not able to make the requests to create posts the "create-record"[1] endpoint which the docs say is a PDS endpoint, results in the following error: "OAuth tokens are meant for PDS access only". Which seems a bit baffling to me as this is a PDS endpoint I'm trying to access


[1]: docs.bsky.app/docs/api/com-atproto-repo-create-record

**Comment by yamarten** [comment]
I'm glad that the problem seems to have progressed.

Regarding the new error, I suspect that you are trying to connect to the entryway (bsky.social) again instead of your PDS (hollowfoot.us-west.host.bsky.network).
I can't find the error message in the current PDS code.

**Comment by bobbyzholmes** [comment]
Across the line @yamarten 

Thank you for the help the issues I faced and the resolution:

1. Using `Authorization: Bearer<token>` the correct value should have been `Authorization: DPoP <token>` thank you Dan-Ratner over at Stack OverFlow[1]
2. When making requests to PDS (/xrpc/com.atproto.repo.createRecord), I was using the entryway (bsky.social) instead of the PDS endpoint. The correct endpoint was extracted from the JWT > "aud" thank you @yamarten 
3. The final error "message":"DPoP nonce mismatch", I was getting when making PDS requests, was due to the dpop nonce changing/expiring, and I hadn't dealt with the change/reply from requests resulting in 401 errors. 

[1] https://stackoverflow.com/questions/79529947/bluesky-social-oauth-scope-bad-token-scope/79554470#79554470
